### ISSUE TITLE
Fix "zephir install" command to not require to be run as root and other broken things

### ISSUE THREAD

**Issue by dschissler** [zephir-lang/zephir#1733]
Currently zephir commands fail if it needs root access like for example to install the extension.  This can be achieved by adding `sudo` to only the specific shell outs that require this access.  Then the command can be run as a normal user and it will prompt them for sudo.  Maybe PHP `passthru` is the best way to do this, I'm not sure.

Also using sudo only on the specific commands can help project the system from a misconfigured program.  For example a make does a lot of stuff but it can't harm the system if it doesn't have root and then the install part simply copies the file to the root permission location.  This is a long established convention for development tools.

**Comment by dschissler** [comment]
@sergeyklay the latest `0.11.4-eb5db1a` works perfectly in this regard for `zephir build` in that it asks for sudo only at the appropriate location.  However `zephir compile` and then `zephir install` don't work as I expect them to.  What is the intended behavior of `compile` and `install`?  I expected `compile` to only compile the zephir -> C -> binary without installing it.  Then I expected `install` to not try to compile the source and only to install the extension.  As it is now the `install` subcommand requires root to be used in its entirety but `build` is working in a more intelligent manner.

Perhaps the problem is only with `zephir install`?  I'm thinking that it should say:

```
install    Installs the compiled extension in the extension directory
```

Notice that I added "compiled" and removed the disclaimer about root.  The sudo prompt takes care of this and also you aren't currently advertising this for `build` either.

I don't think that `install` should attempt to compile the source if it has not been compiled.  Actually I don't think that it should even check that the extension is up to date.  IMO just keep it simple and install the extension.  So really that is just short hand for a few shell commands.  Otherwise you have essentially `build` and `install` doing the same thing but with `install` being more primitive and broken in regards to sudo.

So perhaps everything is good and you should only simplify/rework the `install` subcommand.

So in my mind `zephir build` really does the following:

```
zephir generate
zephir compile (also generates)
zephir install
```

**Comment by dschissler** [comment]
I think that to be completely consistent that:

* `install` shouldn't try to `generate` or `build`,
* compile shouldn't try to `generate`

These commands can be strung together manually for intense customization and debugging and 95% of Zephir developers just will use `zephir build`.  Again, I can't say that its common sense now and the commands try to do too much, more than they advertise.

You could even put a small help on the zephir build description (start here), (use this) or something like that since its non-obvious on first use which command to use to just build and install the damn extension.  I really like how you've removed the poop in the Zephir setup in this 0.11.4 and if you clean up the commands themselves a bit then I can say that Zephir is user friendly.  As it is now it just falls a little bit short.

**Comment by dschissler** [comment]
In fact I'd like you to add "bug" to this instead of just "enhancement".  To use `zephir install` I need to call it as root `sudo zephir install` and then my ext directory is owned by root and so I can't work with it as my normal user.  That is just all broken to hell.

**Comment by sergeyklay** [comment]
You are reading my mind. A few days ago, I thought that it would be good to change this.

That's what these commands actually do:
- `generate`: Generates the C sources from Zephir without compiling them
- `compile`: Compiles the extension without installing it (always calls `generate` command)
- `install`: Compiles and installs the extension (always calls `compile` command)
- `build`: Calls the `install` command

I want to change this so that it will works in such way

- `generate`: Generates the C sources from Zephir without compiling them
- `compile`: Compiles the extension without installing it (if there is no generated code will raise an error)
- `install`: Only installs the extension (if there is no compiled code will raise an error)
- `build`: Should call `generate`, `compile` then `install`

---

BTW, I don't think this is related to this issue's topic  :)

**Comment by dschissler** [comment]
I changed the title.  The "and other broken things" covers it.

**Comment by Jeckerson** [comment]
Fixed in `0.13.0`
